docs/synchronization.md

```diff
@@ -0,0 +1,220 @@
+// hash() is a custom method  hashing the current file.
+```
+
+### konnector
+
+The konnector attribute is used as a key reference to retrieve all document synchronized by this konnector.
+
+For now, the `konnector` option will be needed for every call. But it could be interesting to initialize some kind of runner with it, which may encapsulate all the `cozy-konnector-libs` functions.
+
+### shouldSave
+
+The shouldSave function returns a boolean which indicate if the entry should be saved. The default method should be:
+```js
+async function shouldSave (entry, existingDocument, cozy) => {
+  return Promise.resolve(!existingDocument)
+}
+```
+When calling this function, it is assumed that `cozy-konnector-libs` first queries the database for an exisring document, based on synchronization data and pass the document resulting the query to the `shouldSave` function, event if this document is `null`.
+
+As a third parameter, a cozyClient instance is passed, it may be used by a konnector to perform another query.
+
+Because of this third parameter, the whole `shouldSave` method is asynchronous.
+
+### shouldUpdate
+
+The `shouldUpdate` function acts exactly like `shouldSave`, except that its used to determine if a document should be updated. By default, we make the choice to never update a document (except if it does not have synchronization data, see `findLegacyDocument` above). But some konnectors like `Linxo` related ones need to update existing documents.
+
+The default method is:
+```js
+async function shouldUpdate (entry, existingDocument, cozy) {
+  // Update an existing document only if it does not have synchronization data
+  const hasSyncData = !!existingDocument && !!existingDocument.metadata && !! existingDocument.metadata.sync
+  return Promise.resolve(!hasSyncData)
+}
+```
+
+## Whole example
+
+Here is a whole and naive example of how synchronization data and synchronization options could be used.
+
+```js
+saveFiles(files, {
+  findLegacyDocument: (entry, cozy) => {
+    return cozy.files.statByPath(getEntryPath(entry))
+  },
+  getSyncData: (file) => ({
+    fileName: file.name
+  }),
+  idAttribute: 'uuid',
+  konnector: 'myservice',
+  shouldSave: (file) => {
+    // For whatever reason we only save files created after 2010
+    return Promise.resolve(moment(file.creationDate).year().isAfter(2010))
+  },
+  shouldUpdate: (file, existingDocument) => {
+    const hasBeenModified = moment(file.modificationDate).isAfter(existingDocument.metadata.sync.last_sync)
+    return Promise.resolve(hasBeenModified)
+  }
+})
+```
+Every synchronized document will contain:
+```json
+{
+  "metadata": {
+    "sync": {
+      "konnector": "myservice",
+      "id": "<entry_id>",
+      "last_sync": "<now>",
+      "fileName": "<fileName>"
+    }
+  }
+}
+```
+
+## Remaining questions
+Do we need to pass as second parameter to `shouldSave`, and `shouldDelete` the whole existing document or just the existing synchronzation data ?
```

Comment:
IMHO since all sync data should be unique, I think the sync data must be enough. Except for the legacy doc which will need the entire document.

Reply:
You're right about the legacy document.
Also, if we pass only sync data, we could not be able to detect a difference between a non existing document and an existing document without sync data.

Reply:
I would anyway prefer to pass all the document to let the connector developpers handle some cases we did not think of.